### PR TITLE
Fix data frame initialisation in `bind_rows()`

### LINKED ISSUE #5417: `bind_rows()` does not check list contents
before calling `new_data_frame()`:

```r
dplyr::bind_rows(list(x = 1), list(x = 1, y = 1:2))
#> Error: Internal error in `vec_assign()`: `value` should have been recycled to fit `x`.
```

### PR DESCRIPTION
**Pull request by lionel-**
Closes #5417.

### CODE REVIEW

**Comment by lionel-** [comment]
Performance impact:

```r
new_list <- purrr::map(1:10000, function(x) list(A = 2 * x, B = 4 * x)
bench::mark(bind_rows(new_list))[1:8]

#> # Master
#>   expression               min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr>          <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 bind_rows(new_list)    181ms    194ms      5.25     821KB     14.0     3     8

#> # This PR
#>   expression               min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc
#>   <bch:expr>          <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>
#> 1 bind_rows(new_list)    552ms    552ms      1.81     821KB     9.05     1     5
```

File: R/bind.r [hunk 1]
```diff
@@ -128,7 +128,7 @@ bind_rows <- function(..., .id = NULL) {
   first <- dots[[1L]]
   dots <- map(dots, function(.x) {
     if (vec_is_list(.x)) {
-      .x <- new_data_frame(as.list(.x))
+      .x <- vctrs::data_frame(!!!.x)
```
Comment (romainfrancois) [inline R/bind.r:131]: Thanks. Do we envision an argument in `vec_rbind()` so that we would not have to do that conversion ?

**Review by romainfrancois — APPROVED** [review]

File: R/bind.r (hunk 1 above)
Comment (lionel-) [reply R/bind.r:131]: We purposely don't interpret lists as data frames in `vec_rbind()`. They are interpreted as rows like atomic vectors.

Comment (lionel-) [reply R/bind.r:131]: This wouldn't speed up initialisation anyway. Hopefully we can make `df_list()` faster in the future...

**Comment by hadley** [comment]
And #5749

**Comment by lionel-** [comment]
@DavisVaughan Good point about name-repair. We don't need to repair names before calling `vec_rbind()` since `vec_rbind()` repairs them. Fixed in https://github.com/tidyverse/dplyr/commit/827615c591da2f77a24acc5dfcc0a2ec5d858ccd.

The perf example above was reduced from 370ms to 240ms. Thanks!

**Pull request merged**